Unknown tool hallucination: masc_claim_task

## Symptom
An agent called a tool named `masc_claim_task` which does not exist in the MASC MCP server.

```
[ERROR] tool call failed: masc_claim_task — detail=Unknown tool: masc_claim_task
```

## Affected sequences
- seq=55937

## Impact
The agent receives a tool-not-found error and must retry or fall back. This wastes a turn and may confuse the agent's planning.

## Possible causes
1. **LLM hallucination**: The model invented `masc_claim_task` based on the pattern `masc_claim_next` + `masc_add_task`.
2. **Stale documentation**: If any prompt or instruction document mentions `masc_claim_task`, the LLM may be copying it verbatim.
3. **Tool list drift**: The available tools changed and the agent's cached tool list is stale.

## Notes
- The correct tool for claiming a task is `masc_claim_next`.
- The correct tool for adding a task is `masc_add_task`.
- Search for `masc_claim_task` in docs, prompts, and keeper configs to rule out documentation drift.

--- a/lib/mcp_server_eio_execute.ml
+++ b/lib/mcp_server_eio_execute.ml
@@ -552,13 +552,25 @@ let execute_tool_eio ~sw ~clock ?mcp_session_id ?auth_token state ~name ~argumen
         Tool_inline_dispatch.dispatch inline_ctx ~name
   in
 
+  (* #9784: enrich Unknown tool errors with closest-name suggestions so the
+     LLM can self-correct on the next turn rather than re-emit the same
+     hallucinated name. Suggestions come from a similarity scan of the
+     full tool registry. *)
+  let format_unknown_tool_error ~reason =
+    let suggestions = Tool_dispatch.find_similar_names ~query:name () in
+    match suggestions with
+    | [] -> Printf.sprintf "Unknown tool: %s (%s)" name reason
+    | xs ->
+        Printf.sprintf "Unknown tool: %s — did you mean: %s? (%s)"
+          name (String.concat ", " xs) reason
+  in
   (* Primary dispatch: mint token at I/O boundary, then O(1) tag lookup.
      Tool_token validates the name exists in the tag registry (Parse, Don't
      Validate). If mint fails, the tool is truly unknown. *)
   match Tool_dispatch.mint_token ~name with
   | Error reason ->
       with_system_internal_audit ~agent_name
-        (false, Printf.sprintf "Unknown tool: %s (%s)" name reason)
+        (false, format_unknown_tool_error ~reason)
   | Ok _token ->
       (* Token proves the name is registered in at least one registry.
          lookup_tag None after mint is a registry inconsistency (tool in

--- a/lib/tool_dispatch.ml
+++ b/lib/tool_dispatch.ml
@@ -242,3 +242,40 @@ let mint_token ~name =
     Tool_token.mint_with
       ~validate:(fun n -> Hashtbl.mem tag_registry n || Hashtbl.mem registry n)
       ~name)
+
+(** Enumerate every tool name registered in either the tag_registry (primary)
+    or the handler registry (fallback). Used by [find_similar_names] to
+    drive "did you mean" suggestions for Unknown tool errors (#9784). *)
+let all_registered_names () =
+  with_dispatch_ro (fun () ->
+    let acc =
+      Hashtbl.fold (fun n _ a -> n :: a) tag_registry []
+    in
+    Hashtbl.fold
+      (fun n _ a -> if List.mem n a then a else n :: a)
+      registry acc)
+
+(* #9784: Unknown tool errors must include closest-name suggestions so the
+   LLM can self-correct on the next turn. Jaccard works well for snake_case
+   tool names because Text_similarity tokenizes on non-alphanumeric and
+   captures shared morphemes via byte n-grams. The default min_score 0.4
+   excludes unrelated names while accepting near-misses like
+   masc_claim_task -> masc_claim_next (Jaccard >= 0.5). *)
+let find_similar_names ?(limit = 3) ?(min_score = 0.4) ~query () =
+  let candidates = all_registered_names () in
+  let scored =
+    List.filter_map
+      (fun n ->
+        let s = Text_similarity.jaccard_similarity query n in
+        if s >= min_score then Some (s, n) else None)
+      candidates
+  in
+  let sorted =
+    List.sort (fun (a, _) (b, _) -> Float.compare b a) scored
+  in
+  let rec take k = function
+    | _ when k <= 0 -> []
+    | [] -> []
+    | (_, n) :: rest -> n :: take (k - 1) rest
+  in
+  take limit sorted

--- a/lib/tool_dispatch.mli
+++ b/lib/tool_dispatch.mli
@@ -138,3 +138,16 @@ val tag_registry_count : unit -> int
 
 val mark_tag_registry_initialized : unit -> unit
 val is_tag_registry_initialized : unit -> bool
+
+(** {1 Did-you-mean Suggestions (#9784)} *)
+
+val all_registered_names : unit -> string list
+(** Every tool name registered in either the tag_registry or the
+    handler registry, deduplicated. Iteration order is unspecified. *)
+
+val find_similar_names :
+  ?limit:int -> ?min_score:float -> query:string -> unit -> string list
+(** Return up to [limit] (default 3) tool names from the registries with
+    [Text_similarity.jaccard_similarity] to [query] >= [min_score]
+    (default 0.4), sorted by similarity descending. Used to enrich
+    Unknown tool errors with self-correction hints for LLM clients. *)

--- a/test/test_tool_dispatch.ml
+++ b/test/test_tool_dispatch.ml
@@ -198,4 +198,42 @@ let () =
               check bool "contains error info" true
                 (String.length msg > 0 && Astring.String.is_infix ~affix:"boom" msg));
         ] );
+      ( "did_you_mean_9784",
+        [
+          test_case "find_similar_names returns close match" `Quick (fun () ->
+              register_full ~tool_name:"__sim_masc_claim_next" ~handler:echo_handler;
+              register_full ~tool_name:"__sim_masc_add_task" ~handler:echo_handler;
+              register_full ~tool_name:"__sim_masc_join" ~handler:echo_handler;
+              let suggestions =
+                Tool_dispatch.find_similar_names
+                  ~query:"__sim_masc_claim_task" ()
+              in
+              check bool "non-empty suggestions" true
+                (List.length suggestions >= 1);
+              check bool "top suggestion is closest" true
+                (List.hd suggestions = "__sim_masc_claim_next"));
+          test_case "find_similar_names empty when nothing close" `Quick (fun () ->
+              let suggestions =
+                Tool_dispatch.find_similar_names
+                  ~query:"completely_different_xyzqq_unrelated" ()
+              in
+              check int "no suggestions" 0 (List.length suggestions));
+          test_case "find_similar_names respects limit" `Quick (fun () ->
+              for i = 1 to 5 do
+                register_full
+                  ~tool_name:(Printf.sprintf "__limit_test_tool_%d" i)
+                  ~handler:echo_handler
+              done;
+              let suggestions =
+                Tool_dispatch.find_similar_names ~limit:2
+                  ~query:"__limit_test_tool_1" ()
+              in
+              check bool "at most 2 returned" true
+                (List.length suggestions <= 2));
+          test_case "all_registered_names enumerates registry" `Quick (fun () ->
+              register_full ~tool_name:"__enum_check_xyz" ~handler:echo_handler;
+              let all = Tool_dispatch.all_registered_names () in
+              check bool "contains registered name" true
+                (List.mem "__enum_check_xyz" all));
+        ] );
     ]

